contracts/ChainlinkPriceFeedV3.sol

```diff
@@ -0,0 +1,213 @@
+import { IChainlinkPriceFeed } from "./interface/IChainlinkPriceFeed.sol";
+import { IChainlinkPriceFeedV3 } from "./interface/IChainlinkPriceFeedV3.sol";
+import { BlockContext } from "./base/BlockContext.sol";
+import { CachedTwap } from "./twap/CachedTwap.sol";
+
+contract ChainlinkPriceFeedV3 is IChainlinkPriceFeedV3, BlockContext, CachedTwap {
+    using SafeMath for uint256;
+    using Address for address;
+
+    //
+    // STATE
+    //
+
+    uint24 private constant _ONE_HUNDRED_PERCENT_RATIO = 1e6;
+    uint8 internal immutable _decimals;
+    uint24 internal immutable _maxOutlierDeviationRatio;
+    uint256 internal immutable _outlierCoolDownPeriod;
+    uint256 internal immutable _timeout;
+    uint256 internal _lastValidPrice;
+    uint256 internal _lastValidTimestamp;
+    AggregatorV3Interface internal immutable _aggregator;
+
+    //
+    // EXTERNAL NON-VIEW
+    //
+
+    constructor(
+        AggregatorV3Interface aggregator,
+        uint256 timeout,
+        uint24 maxOutlierDeviationRatio,
+        uint256 outlierCoolDownPeriod,
+        uint80 twapInterval
+    ) CachedTwap(twapInterval) {
+        // CPF_ANC: Aggregator is not contract
+        require(address(aggregator).isContract(), "CPF_ANC");
+        _aggregator = aggregator;
+
+        // CPF_IMODR: Invalid maxOutlierDeviationRatio
+        require(maxOutlierDeviationRatio < _ONE_HUNDRED_PERCENT_RATIO, "CPF_IMODR");
+        _maxOutlierDeviationRatio = maxOutlierDeviationRatio;
+
+        _outlierCoolDownPeriod = outlierCoolDownPeriod;
+        _timeout = timeout;
+        _decimals = aggregator.decimals();
+    }
+
+    /// @inheritdoc IChainlinkPriceFeedV3
+    function cacheTwap(uint256 interval) external override {
+        _cachePrice();
+
+        if (interval != 0) {
+            _cacheTwap(interval, _lastValidPrice, _lastValidTimestamp);
+        }
+    }
+
+    //
+    // EXTERNAL VIEW
+    //
+
+    function getLastValidPrice() external view override returns (uint256) {
+        return _lastValidPrice;
+    }
+
+    function getLastValidTimestamp() external view override returns (uint256) {
+        return _lastValidTimestamp;
+    }
+
+    /// @inheritdoc IChainlinkPriceFeedV3
+    function getCachedTwap(uint256 interval) external view override returns (uint256) {
+        (uint256 latestValidPrice, uint256 latestValidTime) = _getCachePrice();
+
+        if (interval == 0) {
+            return latestValidPrice;
+        }
+
+        return _getCachedTwap(interval, latestValidPrice, latestValidTime);
+    }
+
+    function getAggregator() external view override returns (address) {
+        return address(_aggregator);
+    }
+
+    function decimals() external view override returns (uint8) {
+        return _decimals;
+    }
+
+    function isTimedOut() external view override returns (bool) {
+        return _lastValidTimestamp.add(_timeout) > _blockTimestamp();
```

Comment:
Should be `<`?

```solidity
return _lastValidTimestamp.add(_timeout) < _blockTimestamp();

timeout: 10
current timestamp: 100
last timestamp: 80
80 + 10 < 100 => true, time out

timeout: 10
current timestamp: 100
last timestamp: 90
90 + 10 < 100 => false, not time out

timeout: 10
current timestamp: 100
last timestamp: 100
100 + 10 < 100 => false, not time out
```

Reply:
Yes, it's a huge mistake. I have fixed it in https://github.com/perpetual-protocol/perp-oracle-contract/pull/52. Will be merged to here soon.

Reply:
Already merged the https://github.com/perpetual-protocol/perp-oracle-contract/pull/52 into this branch.
cc @vinta @twxia